src/commands/wp.ts

```diff
@@ -0,0 +1,190 @@
+		inputToken: string;
+		sshAuthentication: {
+			host: string;
+			port: number;
+			username: string;
+			privateKey: string;
+			passphrase: string;
+		};
+		command: {
+			guid: string;
+		};
+	};
+}
+
+const TRIGGER_WP_CLI_COMMAND_MUTATION = gql`
+	mutation TriggerWPCLICommandMutation($input: AppEnvironmentTriggerWPCLICommandInput) {
+		triggerWPCLICommandOnAppEnvironment(input: $input) {
+			inputToken
+			sshAuthentication {
+				host
+				port
+				username
+				privateKey
+				passphrase
+			}
+			command {
+				guid
+			}
+		}
+	}
+`;
+
+const getSSHAuthForCommand = async ( appId: number, envId: number, command: string ) => {
+	const api = API();
+
+	return api.mutate( {
+		mutation: TRIGGER_WP_CLI_COMMAND_MUTATION,
+		variables: {
+			input: {
+				id: appId,
+				environmentId: envId,
+				command,
+			},
+		},
+	} ) as Promise< { data: TriggerWPCLICommandMutationResponse } >;
+};
+
+export class WPCliCommandOverSSH {
+	private app: App;
+	private env: AppEnvironment;
+	private track: CommandTracker;
+
+	constructor( app: App, env: AppEnvironment ) {
+		this.app = app;
+		this.env = env;
+
+		this.track = makeCommandTracker( 'wp', {
+			app: this.app.id,
+			env: this.env.id,
+			executionType: 'ssh',
+		} );
+	}
+
+	public async run( command: string ): Promise< void > {
+		if ( ! this.app.id || ! this.env.id ) {
+			console.error( 'No app ID or environment ID provided' );
+
+			await this.track( 'error', {
+				error: 'no_app_env_id',
+				message: 'No app or env ID provided',
+			} );
+			return;
+		}
+
+		debug( "Requesting SSH authentication for command '%s'", command );
+
+		const sshAuth = await getSSHAuthForCommand( this.app.id, this.env.id, command );
+
+		const data = sshAuth.data?.triggerWPCLICommandOnAppEnvironment;
+
+		debug( 'Connecting to SSH' );
+
+		try {
+			await this.executeCommandOverSSH( {
+				host: data.sshAuthentication.host,
+				port: data.sshAuthentication.port,
+				username: data.sshAuthentication.username,
+				privateKey: data.sshAuthentication.privateKey,
+				passphrase: data.sshAuthentication.passphrase,
+				guid: data.command.guid,
+				inputToken: data.inputToken,
+			} );
+
+			await this.track( 'success', { guid: data?.command.guid } );
+		} catch ( err ) {
+			console.log( err );
+			await this.track( 'error', {
+				guid: data?.command.guid,
+				error: 'ssh_command_failed',
+				message: 'Error executing command over SSH',
+			} );
+		}
+	}
+
+	private async executeCommandOverSSH( {
+		host,
+		port,
+		username,
+		privateKey,
+		passphrase,
+		guid,
+		inputToken,
+	}: {
+		host: string;
+		port: number;
+		username: string;
+		privateKey: string;
+		passphrase: string;
+		guid: string;
+		inputToken: string;
+	} ) {
+		return new Promise( ( resolve, reject ) => {
+			const conn = new ssh2.Client();
+
+			conn
+				.on( 'ready', () => {
+					conn.exec(
+						`GUID=${ guid } INPUT_TOKEN=${ inputToken } VERSION=${ pkg.version }`,
+						( err, stream ) => {
+							if ( err ) throw err;
```

Comment:
```suggestion
							if ( err ) {
							   reject( err );
							   return;
							 }
```
Would this be better? I don't think `throw err` here would cause the promise to reject but I might be wrong, and ssh2 passes the thrown error to `on( 'error' )`

If ssh2 passes the error to `on( 'error' )`, maybe it's worth adding a comment here as it's not obvious, at least for me. If ssh2 does pass the error to `on( 'error' )`, I wonder if all libraries that rely on event works that way - it's a nice way to ensure all errors are handled in one place.

Reply:
Good catch!

100%!